src/mongoose_c2s_presence.erl

```diff
@@ -0,0 +1,452 @@
+    I = gb_sets:del_element(LTo, StateData#roster_state.pres_i),
+    A = gb_sets:del_element(LTo, StateData#roster_state.pres_a),
+    {new_state, StateData#roster_state{pres_i = I, pres_a = A}};
+set_presence_state({invisible, LTo}, StateData, _C2SState) ->
+    I = gb_sets:add_element(LTo, StateData#roster_state.pres_i),
+    A = gb_sets:del_element(LTo, StateData#roster_state.pres_a),
+    {new_state, StateData#roster_state{pres_i = I, pres_a = A}};
+set_presence_state({available, LTo}, StateData, _C2SState) ->
+    I = gb_sets:del_element(LTo, StateData#roster_state.pres_i),
+    A = gb_sets:add_element(LTo, StateData#roster_state.pres_a),
+    {new_state, StateData#roster_state{pres_i = I, pres_a = A}}.
+
+is_present_or_invisible(StateData) ->
+    S = ejabberd_c2s_state:get_handler_state(mod_roster, StateData),
+    case S#roster_state.pres_last of
+        undefined ->
+            S#roster_state.pres_invis;
+        _ ->
+            true
+    end.
+
+lowcase_and_bare(JID) ->
+    LJID = jid:to_lower(JID),
+    {LJID, jid:to_bare(LJID)}.
+
+get_subscriptions(to, #roster_state{pres_t = T}) -> T;
+get_subscriptions(from, #roster_state{pres_f = T}) -> T;
+get_subscriptions(available, #roster_state{pres_a = T}) -> T;
+get_subscriptions(invisible, #roster_state{pres_i = T}) -> T;
+get_subscriptions(last, #roster_state{pres_last = T}) -> T;
+get_subscriptions(trusted, S) ->
+    #roster_state{pres_f = F, pres_a = A} = ejabberd_c2s_state:get_handler_state(mod_roster, S),
+    lists:filter(fun(JID) -> gb_sets:is_element(JID, F) end,
+                 gb_sets:to_list(A));
+get_subscriptions(from_except_self, S) ->
+    JID = ejabberd_c2s_state:jid(S),
+    SelfJID = jid:to_lower(jid:to_bare(JID)),
+    Froms = get_subscriptions(from, ejabberd_c2s_state:get_handler_state(mod_roster, S)),
+    gb_sets:to_list(gb_sets:del_element(SelfJID, Froms));
+get_subscriptions(Which, S) ->
+    gb_sets:to_list(
+        get_subscriptions(Which,
+                          ejabberd_c2s_state:get_handler_state(mod_roster, S))).
+
+get_last_presence(State) ->
+    S = ejabberd_c2s_state:get_handler_state(mod_roster, State),
+    S#roster_state.pres_last.
+
+get_presence_timestamp(State) ->
+    S = ejabberd_c2s_state:get_handler_state(mod_roster, State),
+    S#roster_state.pres_timestamp.
+
+get_priority(State) ->
+    case get_last_presence(State) of
+        undefined ->
+            0;
+        OldPresence ->
+            get_priority_from_presence(OldPresence)
+    end.
+
+presence_update_to_available(Packet, StateData, _C2SState) ->
+    {new_state, StateData#roster_state{pres_last = Packet,
+                                       pres_invis = false,
+                                       pres_timestamp = calendar:universal_time()}}.
+
+is_unavailable(C2SState) ->
+    S = ejabberd_c2s_state:get_handler_state(mod_roster, C2SState),
+    (S#roster_state.pres_last == undefined) or S#roster_state.pres_invis.
```

Comment:
I don't see any reason for `or` instead of `orelse` here (I know it's moved from `ejabberd_c2s`)